### PR TITLE
Add ability for dropdown to be shown at the beginning

### PR DESCRIPTION
**Pull request by thangdinh**
There are some use cases when we want to have dropdown shown by default, it's not currently supported. 
Tests are added.
@Addepar/fire

### CODE REVIEW

File: src/select.coffee [hunk 1]
```diff
@@ -349,11 +368,19 @@ Ember.AddeparMixins.ResizeHandlerMixin, Ember.Widgets.KeyboardHelper,
...
+  * item does not exist.
+  * @function
+  ###
+  updateHighlightedItem: ->
     if @get('showDropdown')
       highlighted = @get('highlighted')
       if not @get('selectableOptions').contains(highlighted)
         @set 'highlighted', @get('selectableOptions.firstObject')
+
+  selectableOptionsDidChange: Ember.observer ->
```
Comment (philpee2) [inline src/select.coffee]: :cry:

File: tests/unit/select_test.coffee [hunk 2]
```diff
@@ -116,6 +116,62 @@ test 'Test keyboard interaction', ->
...
+  expect 3
+
+  select = @subject
+    content: ['foo', 'bar', 'barca', 'baz']
+  @append()
+
+  selectComponent = select.$()
+
+  validateDropdownVisible = (messageVisible) ->
+    ok isVisible(find '.ember-select-results', selectComponent),
```
Comment (philpee2) [inline tests/unit/select_test.coffee]: Is `find` needed here?

**Comment by thangdinh** [comment]
retest

File: src/select.coffee [hunk 3]
```diff
@@ -80,8 +81,19 @@ Ember.AddeparMixins.ResizeHandlerMixin, Ember.Widgets.KeyboardHelper,
...
+  * @type {Boolean}
+  ###
   showDropdown: no
 
+  ###*
+  * Flag indicating that dropdown will always appear as part of the select
+  * component
+  * @type {Boolean}
+  ###
+  lockDropdownOnShow: no
```
Comment (jordansmith42) [inline src/select.coffee]: Can we prefix this with `is` or `should`? Kind of a weird name too, shouldInitWithDropdown or something along those lines?

File: src/select.coffee (hunk 1 above)
Comment (jordansmith42) [inline src/select.coffee]: Why is this necessary?

File: src/select.coffee [hunk 4]
```diff
@@ -492,8 +531,10 @@ Ember.AddeparMixins.ResizeHandlerMixin, Ember.Widgets.KeyboardHelper,
 
   actions:
     toggleDropdown: (event) ->
-      return if @get('disabled')
-      @toggleProperty 'showDropdown'
+      if @get 'showDropdown'
+        @send 'hideDropdown' if not @get('lockDropdownOnShow')
```
Comment (jordansmith42) [inline src/select.coffee]: Why handle this differently than other cases of showDropdown/lockDropdownOnShow?

File: src/select.coffee (hunk 3 above)
Comment (embooglement) [inline src/select.coffee]: I like Jordan's suggestio

File: src/select.coffee [hunk 5]
```diff
@@ -42,11 +43,11 @@ Ember.Widgets.SelectOptionView = Ember.ListItemView.extend
   click: ->
     return if @get('content.isGroupOption')
     @set 'controller.selection', @get('content')
-    @get('controller').userDidSelect @get 'content'
-    # if there's a selection and the dropdown is unexpanded, we want to
-    # propagate the click event
-    # if the dropdown is expanded and we select something, don't propagate
+    # Check if the dropdown is currently shown to make sure that we has clicked
```
Comment (embooglement) [inline src/select.coffee]: Should be "make sure that we have clicked"

File: src/select.coffee [hunk 6]
```diff
@@ -42,11 +43,11 @@ Ember.Widgets.SelectOptionView = Ember.ListItemView.extend
   click: ->
```
Comment (embooglement) [inline src/select.coffee]: Should the comments in this function be included as documentation instead?

File: src/select.coffee [hunk 7]
```diff
@@ -80,8 +81,19 @@ Ember.AddeparMixins.ResizeHandlerMixin, Ember.Widgets.KeyboardHelper,
   # we need to set tabindex so that div responds to key events
   tabindex: 0
 
+  ###*
+  * Flag controlling the dropdown's appearance of the select component
```
Comment (embooglement) [inline src/select.coffee]: This comment is a little confusing to me. I think it should be something like "Flag that determines whether or not the dropdown is shown".

File: src/select.coffee (hunk 3 above)
Comment (thangdinh) [inline src/select.coffee]: It's not `init with Dropdown` but instead, it will never collapse the dropdown. 
So Jordan+Kevin's suggestions will become `shouldLockDropdownOnShow`? And yeah, we have 80 characters length limit.

Comment (embooglement) [inline src/select.coffee]: `shouldLockDropdownOnShow` seems fine to me.

File: src/select.coffee (hunk 1 above)
Comment (thangdinh) [inline src/select.coffee]: 1. I didn't put an attempt to fix all of the observers in this component in this PR. I happened to find some of the observers are straightforward to be remove so I remove them. 
2. The reason of this observer (as part of the original solution) was to tackle the problem that the content of the selection is changed, for which the current highlighted item does not exist anymore. 
It is not straightforward to build computed properties because here we need the highlighted index bound to the UI, while we need the highlighted item for other operations. With only highlighted index, we cannot know if the item is already removed or not. Something like that. I can explain in person when we want to remove this observer.

File: src/select.coffee (hunk 4 above)
Comment (thangdinh) [inline src/select.coffee]: What do you mean? 
This is for the case of toggling the dropdown. Where we have an `else` case to open the dropdown if it's not there yet. Other logics are aiming to hide the dropdown only.
Is it the answer of your question?

File: tests/unit/select_test.coffee (hunk 2 above)
Comment (thangdinh) [inline tests/unit/select_test.coffee]: It was there before we shaped the way we write tests. I will remove all of them when I refactor the tests here.

File: src/select.coffee (hunk 3 above)
Comment (jordansmith42) [inline src/select.coffee]: Works for me

File: src/select.coffee (hunk 4 above)
Comment (jordansmith42) [inline src/select.coffee]: Yep, I got it now

**Comment by thangdinh** [comment]
I decided to close this one again as I have a simpler solution for my problem https://github.com/Addepar/ember-widgets/pull/222.
I will revise this PR as an improvement PR (killing observers, and newly added tests) later.
